File: sdk/csharp/ExampleJWT.cs
/**
 * C2M API C# SDK Example with JWT Authentication
 */

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using C2M.Api.Api;
using C2M.Api.Client;
using C2M.Api.Model;

namespace C2M.Api.Example
{
    public class C2MClient
    {
        private readonly string _clientId;
        private readonly string _clientSecret;
        private readonly string _baseUrl;
        private readonly HttpClient _httpClient;
        private string _longToken;
        private string _shortToken;
        private DateTime _tokenExpiry;

        public C2MClient(string clientId, string clientSecret, string baseUrl = "https://api.c2m.com/v2")
        {
            _clientId = clientId;
            _clientSecret = clientSecret;
            _baseUrl = baseUrl;
            _httpClient = new HttpClient();
        }

        private async Task<string> GetLongTokenAsync()
        {
            var request = new
            {
                grant_type = "client_credentials",
                client_id = _clientId,
                client_secret = _clientSecret,
                scopes = new[] { "jobs:submit", "jobs:read", "templates:read" },
                ttl_seconds = 2592000 // 30 days
            };

            var content = new StringContent(
                JsonConvert.SerializeObject(request),
                Encoding.UTF8,
                "application/json"
            );

            var response = await _httpClient.PostAsync($"{_baseUrl}/auth/tokens/long", content);
            response.EnsureSuccessStatusCode();

            var responseData = JsonConvert.DeserializeObject<JObject>(
                await response.Content.ReadAsStringAsync()
            );

            _longToken = responseData["access_token"].ToString();
            Console.WriteLine($"Long-term token obtained: {responseData["token_id"]}");
            return _longToken;
        }

        private async Task<string> GetShortTokenAsync()
        {
            if (string.IsNullOrEmpty(_longToken))
            {
                await GetLongTokenAsync();
            }

            var request = new
            {
                scopes = new[] { "jobs:submit" }
            };

            var content = new StringContent(
                JsonConvert.SerializeObject(request),
                Encoding.UTF8,
                "application/json"
            );

            _httpClient.DefaultRequestHeaders.Clear();
            _httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {_longToken}");

            var response = await _httpClient.PostAsync($"{_baseUrl}/auth/tokens/short", content);
            response.EnsureSuccessStatusCode();

            var responseData = JsonConvert.DeserializeObject<JObject>(
                await response.Content.ReadAsStringAsync()
            );

            _shortToken = responseData["access_token"].ToString();
            _tokenExpiry = DateTime.Parse(responseData["expires_at"].ToString());
            Console.WriteLine($"Short-term token obtained: {responseData["token_id"]}");
            return _shortToken;
        }

        private async Task<string> EnsureAuthenticatedAsync()
        {
            if (string.IsNullOrEmpty(_shortToken) || DateTime.UtcNow.AddMinutes(1) >= _tokenExpiry)
            {
                await GetShortTokenAsync();
            }
            return _shortToken;
        }

        private async Task<Configuration> CreateConfigurationAsync()
        {
            var token = await EnsureAuthenticatedAsync();
            
            var config = new Configuration();
            config.BasePath = _baseUrl;
            config.AccessToken = token;
            
            return config;
        }

        public async Task<object> SubmitJobAsync(Dictionary<string, object> jobParams)
        {
            var config = await CreateConfigurationAsync();
            var apiInstance = new DefaultApi(config);

            try
            {
                var result = await apiInstance.SubmitSingleDocWithTemplateParamsAsync(jobParams);
                return result;
            }
            catch (ApiException e)
            {
                if (e.ErrorCode == 401 || e.ErrorCode == 403)
                {
                    // Token expired, refresh and retry
                    _shortToken = null;
                    config = await CreateConfigurationAsync();
                    apiInstance = new DefaultApi(config);
                    var result = await apiInstance.SubmitSingleDocWithTemplateParamsAsync(jobParams);
                    return result;
                }
                throw;
            }
        }
    }

    class Program
    {
        static async Task Main(string[] args)
        {
            try
            {
                var client = new C2MClient(
                    "your-client-id",
                    "your-client-secret"
                );

                var jobParams = new Dictionary<string, object>
                {
                    ["templateId"] = "standard-letter",
                    ["documentUrl"] = "https://example.com/document.pdf",
                    ["recipients"] = new[]
                    {
                        new Dictionary<string, object>
                        {
                            ["name"] = "John Doe",
                            ["address"] = new Dictionary<string, object>
                            {
                                ["line1"] = "123 Main St",
                                ["city"] = "New York",
                                ["state"] = "NY",
                                ["zip"] = "10001"
                            }
                        }
                    }
                };

                var result = await client.SubmitJobAsync(jobParams);
                Console.WriteLine($"Job submitted: {JsonConvert.SerializeObject(result)}");
            }
            catch (Exception e)
            {
                Console.WriteLine($"Error: {e.Message}");
            }
        }
    }
}

File: sdk/swift/ExampleJWT.swift
/**
 * C2M API Swift SDK Example with JWT Authentication
 */

import Foundation
import OpenAPIClient

class C2MClient {
    private let clientId: String
    private let clientSecret: String
    private let baseURL: String
    private var longToken: String?
    private var shortToken: String?
    private var tokenExpiry: Date?
    
    init(clientId: String, clientSecret: String, baseURL: String = "https://api.c2m.com/v2") {
        self.clientId = clientId
        self.clientSecret = clientSecret
        self.baseURL = baseURL
    }
    
    private func getLongToken(completion: @escaping (Result<String, Error>) -> Void) {
        let url = URL(string: "\(baseURL)/auth/tokens/long")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "grant_type": "client_credentials",
            "client_id": clientId,
            "client_secret": clientSecret,
            "scopes": ["jobs:submit", "jobs:read", "templates:read"],
            "ttl_seconds": 2592000 // 30 days
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let accessToken = json["access_token"] as? String,
                  let tokenId = json["token_id"] as? String else {
                completion(.failure(NSError(domain: "C2MClient", code: 1, userInfo: [NSLocalizedDescriptionKey: "Failed to parse token response"])))
                return
            }
            
            self?.longToken = accessToken
            print("Long-term token obtained: \(tokenId)")
            completion(.success(accessToken))
        }.resume()
    }
    
    private func getShortToken(completion: @escaping (Result<String, Error>) -> Void) {
        if longToken == nil {
            getLongToken { [weak self] result in
                switch result {
                case .success:
                    self?.getShortToken(completion: completion)
                case .failure(let error):
                    completion(.failure(error))
                }
            }
            return
        }
        
        let url = URL(string: "\(baseURL)/auth/tokens/short")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(longToken!)", forHTTPHeaderField: "Authorization")
        
        let body: [String: Any] = [
            "scopes": ["jobs:submit"]
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let accessToken = json["access_token"] as? String,
                  let tokenId = json["token_id"] as? String,
                  let expiresAt = json["expires_at"] as? String else {
                completion(.failure(NSError(domain: "C2MClient", code: 2, userInfo: [NSLocalizedDescriptionKey: "Failed to parse short token response"])))
                return
            }
            
            let formatter = ISO8601DateFormatter()
            self?.shortToken = accessToken
            self?.tokenExpiry = formatter.date(from: expiresAt)
            print("Short-term token obtained: \(tokenId)")
            completion(.success(accessToken))
        }.resume()
    }
    
    private func ensureAuthenticated(completion: @escaping (Result<String, Error>) -> Void) {
        let now = Date()
        let buffer = now.addingTimeInterval(60) // 1 minute buffer
        
        if shortToken == nil || tokenExpiry == nil || buffer >= tokenExpiry! {
            getShortToken(completion: completion)
        } else {
            completion(.success(shortToken!))
        }
    }
    
    private func configureAPI(token: String) {
        OpenAPIClientAPI.basePath = baseURL
        OpenAPIClientAPI.customHeaders["Authorization"] = "Bearer \(token)"
    }
    
    func submitJob(jobParams: [String: Any], completion: @escaping (Result<Any, Error>) -> Void) {
        ensureAuthenticated { [weak self] result in
            switch result {
            case .success(let token):
                self?.configureAPI(token: token)
                
                // Create job submission request
                // Note: This would need to be adapted based on the actual Swift SDK model classes
                DefaultAPI.submitSingleDocWithTemplateParams(body: jobParams) { response, error in
                    if let error = error as NSError? {
                        if error.code == 401 || error.code == 403 {
                            // Token expired, refresh and retry
                            self?.shortToken = nil
                            self?.submitJob(jobParams: jobParams, completion: completion)
                        } else {
                            completion(.failure(error))
                        }
                    } else if let response = response {
                        completion(.success(response))
                    }
                }
                
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
}

// Usage example
class ExampleUsage {
    static func main() {
        let client = C2MClient(
            clientId: "your-client-id",
            clientSecret: "your-client-secret"
        )
        
        let jobParams: [String: Any] = [
            "templateId": "standard-letter",
            "documentUrl": "https://example.com/document.pdf",
            "recipients": [[
                "name": "John Doe",
                "address": [
                    "line1": "123 Main St",
                    "city": "New York",
                    "state": "NY",
                    "zip": "10001"
                ]
            ]]
        ]
        
        client.submitJob(jobParams: jobParams) { result in
            switch result {
            case .success(let response):
                print("Job submitted: \(response)")
            case .failure(let error):
                print("Error: \(error.localizedDescription)")
            }
        }
    }
}